components/proxy/src/test/java/com/hotels/styx/infrastructure/AbstractRegistryTest.java

```diff
@@ -0,0 +1,202 @@
+
+import static com.hotels.styx.infrastructure.AbstractRegistryTest.IdObject.idObject;
+import static org.mockito.Matchers.any;
+import static org.mockito.Matchers.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+public class AbstractRegistryTest {
+
+    @Test
+    public void notifiesOfInitialChanges() {
+        TestRegistry registry = new TestRegistry();
+        AbstractRegistry.ChangeListener<IdObject> listener1 = mock(AbstractRegistry.ChangeListener.class);
+        AbstractRegistry.ChangeListener<IdObject> listener2 = mock(AbstractRegistry.ChangeListener.class);
+
+        registry.addListener(listener1);
+        registry.addListener(listener2);
+
+        registry.set(ImmutableList.of(idObject("a", "1"), idObject("b", "2")));
+
+        Registry.Changes<IdObject> changes = changeSet()
+                .added(idObject("a", "1"), idObject("b", "2"))
+                .build();
+
+        verify(listener1).onChange(eq(changes));
+        verify(listener2).onChange(eq(changes));
+    }
+
+    @Test
+    public void notifiesNewListenersImmediately() {
+        TestRegistry registry = new TestRegistry();
+        registry.set(ImmutableList.of(idObject("a", "1"), idObject("b", "2")));
+
+        AbstractRegistry.ChangeListener<IdObject> listener1 = mock(AbstractRegistry.ChangeListener.class);
+        registry.addListener(listener1);
+
+        Registry.Changes<IdObject> changes = changeSet()
+                .added(idObject("a", "1"), idObject("b", "2"))
+                .build();
+
+        verify(listener1).onChange(eq(changes));
+    }
+
+    @Test
+    public void notifiesOfModifiedEntries() {
+        TestRegistry registry = new TestRegistry();
+
+        AbstractRegistry.ChangeListener<IdObject> listener1 = mock(AbstractRegistry.ChangeListener.class);
+        registry.addListener(listener1);
+
+        registry.set(ImmutableList.of(idObject("a", "1")));
+
+        registry.set(ImmutableList.of(idObject("a", "2"), idObject("b", "2")));
+        verify(listener1).onChange(eq(changeSet()
+                .added(idObject("b", "2"))
+                .updated(idObject("a", "2"))
+                .build()));
+    }
+
+    @Test
+    public void identifiableIsNotConsideredChangedWhenItsValueIsNotChanged() {
+        TestRegistry registry = new TestRegistry();
+
+        AbstractRegistry.ChangeListener<IdObject> listener1 = mock(AbstractRegistry.ChangeListener.class);
+        registry.addListener(listener1);
+        verify(listener1).onChange(any(Registry.Changes.class));
+
+        registry.set(ImmutableList.of(idObject("a", "1"), idObject("b", "2")));
+        verify(listener1, times(2)).onChange(any(Registry.Changes.class));
+
+        // Does not change:
+        registry.set(ImmutableList.of(idObject("a", "1"), idObject("b", "2")));
+        verify(listener1, times(2)).onChange(any(Registry.Changes.class));
```

Comment:
Perhaps it would be more intuitive to use `verifyNoMoreInteractions(listener1);`